### ISSUE TITLE
static call to stateful function reverts with no bytes

### ISSUE THREAD

**Issue by mshakeg** [hashgraph/hedera-json-rpc-relay#517]
### Description

Making a static call to simulate the result of a stateful function call reverts with no bytes i.e. `bytes.length == 0`, however making a static call to simulate a view/pure function call reverts with the expected bytes.

### Steps to reproduce

1. In a hardhat project save the following contracts to `SimpleStorage.sol`:
```solidity
// SPDX-License-Identifier: GPL-3.0
pragma solidity >=0.7.0 <0.9.0;

contract SimpleCaller {

    // call this function to simulate static call
    function staticChange(address ssAddress, uint256 _number) public returns (uint256 a)
    {
        try SimpleStorage(ssAddress).updateNumber(_number) {} catch (bytes memory reason) {
            if (reason.length == 0) { // condition is true when SimpleStorage.updateNumber function is stateful
                return (404);
            }

            a = abi.decode(reason, (uint256));
        }
    }

    // this function is called by SimpleStorage.updateNumber and reverts with the simulated number
    function updateNumberCallback(uint256 _number) public pure {
        assembly {
            let ptr := mload(0x20)
            mstore(ptr, _number)
            revert(ptr, 32)
        }
    }
}

contract SimpleStorage {
    uint256 public number;
    uint256 public square;

    function updateNumber(uint256 _number) external returns (uint256) {

        square = _number ** 2; // state changing statement; callStatic works when this line is commented out

        SimpleCaller(msg.sender).updateNumberCallback(_number);
        return number;
    }

    function setNumber(uint256 _number) external returns (uint256) {
        number = _number;

        return number;
    }
}
```
2. Save the following test script to `./test/simple.ts`:
```ts
import { ethers } from "hardhat";

import { ContractFactory } from "@ethersproject/contracts";

import { deployOverrides } from "./utilities";

import { SimpleCaller, SimpleStorage } from "../types";

let _simpleStorage: SimpleStorage;
let _simpleCaller: SimpleCaller;

const deploySimple = async (SimpleStorage: ContractFactory, SimpleCaller: ContractFactory) => {
  const simpleStorage = (await SimpleStorage.deploy(deployOverrides)) as SimpleStorage;

  const simpleStorageAddress = (await simpleStorage.deployTransaction.wait()).contractAddress;

  _simpleStorage = SimpleStorage.attach(simpleStorageAddress) as SimpleStorage;

  console.log("ds 1");

  const simpleCaller = (await SimpleCaller.deploy(deployOverrides)) as SimpleCaller;

  const simpleCallerAddress = (await simpleCaller.deployTransaction.wait()).contractAddress;

  _simpleCaller = SimpleCaller.attach(simpleCallerAddress) as SimpleCaller;

  console.log("ds 2");
};

describe("SimpleStorage", function () {
  before(async () => {
    console.log("s 1");

    const SimpleStorage = await ethers.getContractFactory("SimpleStorage");
    const SimpleCaller = await ethers.getContractFactory("SimpleCaller");

    console.log("s 2");

    await deploySimple(SimpleStorage, SimpleCaller);
  });

  it("should be able to make static call", async function () {
    console.log("t 1");

    const res = await _simpleCaller.callStatic.staticChange(_simpleStorage.address, 101); // STATIC CALL

    console.log("static res:", res);
  });
});
```
3. Install and import the `hardhat-ethers` plugin to `hardhat.config.ts`
4. Run the test: `hardhat test test/simple.ts` and observe the `bytes.length == 0` return value i.e. 404(0x0194) from `callStatic`
5. Comment out  the stateful statement `square = _number ** 2;`  in `SimpleStorage.sol` and observe the expected return value from the `callStatic` i.e. 101(0x65)
6. Confirm the issue is with the hedera node by spinning up a local ganache blockchain and running the test against it, configure the ganache local node in `hardhat.config.ts` networks and change `defaultNetwork` to `ganache` and observe the expected return value 101(0x65):
```
const config: HardhatUserConfig = {
  ...,
  defaultNetwork: "ganache",
  networks: {
    hedera: {
      url: "http://127.0.0.1:7546",
      accounts: [
        "0x105d050185ccb907fba04dd92d8de9e32c18305e097ab41dadda21489a211524",
        "0x2e1d968b041d84dd120a5860cee60cd83f9374ef527ca86996317ada3d0d03e7",
        "0x45a5a7108a18dd5013cf2d5857a28144beadc9c70b3bdbd914e38df4e804b8d8",
        "0x6e9d61a325be3f6675cf8b7676c70e4a004d2308e3e182370a41f5653d52c6bd",
        "0x0b58b1bd44469ac9f813b5aeaf6213ddaea26720f0b2f133d08b6f234130a64f",
        "0x95eac372e0f0df3b43740fa780e62458b2d2cc32d6a440877f1cc2a9ad0c35cc",
        "0x6c6e6727b40c8d4b616ab0d26af357af09337299f09c66704146e14236972106",
        "0x5072e7aa1b03f531b4731a32a021f6a5d20d5ddc4e55acbb71ae202fc6f3a26d",
        "0x60fe891f13824a2c1da20fb6a14e28fa353421191069ba6b6d09dd6c29b90eff",
        "0xeae4e00ece872dd14fb6dc7a04f390563c7d69d16326f2a703ec8e0934060cc7",
      ],
    },
    ganache: {
      url: "http://127.0.0.1:8545",
      chainId: 1337,
      accounts: [
        "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80",
        "0x59c6995e998f97a5a0044966f0945389dc9e86dae88c7a8412f4603b6b78690d",
        "0x5de4111afa1a4b94908f83103eb1f1706367c2e68ca870fc3fb9a804cdab365a",
        "0x7c852118294e51e653712a81e05800f419141751be58f605c371e15141b007a6",
        "0x47e179ec197488593b187f80a00eb0da91f1b9d0b13f8733639f19c30a34926a",
        "0x8b3a350cf5c34c9194ca85829a2df0ec3153be0318b5e2d3348e872092edffba",
        "0x92db14e403b83dfe3df233f83dfa3a0d7096f21ca9b0d6d6b8d88b2b4ec1564e",
        "0x4bbbf85ce3377467afe5d46f804f221813b2bb87f24d81f60f1fcdbf7cbf4356",
        "0xdbda1821b80551c9d65939329250298aa3472ba22feea921c0cf5d620ea67b97",
        "0x2a871d0798f97d79848a013d4936a73bf4cc922c825d33c1cf7073dff6d409c6",
      ],
    },
  },
  ...
};
```

### Hedera network

other

### Version

@hashgraph/hedera-local@1.2.0

### Operating system

macOS

**Comment by mshakeg** [comment]
Hey @Nana-EC , I just wanted to ask if you might have an estimate on when this can be completed. It's core functionality for a component we're working on.

**Comment by Nana-EC** [comment]
Hey @mshakeg we'll triage the issue and determine the scope of the issue and what a potential fix looks like.
I can't speak to a timeline estimate until then.

**Comment by mshakeg** [comment]
Hey @Nana-EC thanks for the reply. Looking forward to hearing back from you.

**Comment by shemnon** [comment]
In Hedera and hashio for `eth_call` we enforce "static" semantics from the very first frame on `eth_call`.  Hence you should only be calling `view` and `pure` functions from `eth_call`.

The commented out line invokes a `SSTORE` operation, which triggers a revert when called in a static context.

We may adjust it when we shift `eth_call` to be invoked in hashio, but in the interim all `eth_call` calls will be static.

**Comment by mshakeg** [comment]
@shemnon thanks for providing context. My understanding was that Hedera's relay was meant to have parity with Ethereum relays(Infura, etc) which do allow static calls to stateful functions to simulate outputs. If this functionality isn't going to be implemented on Hedera in the near future that will significantly impact development as many major DApps require this functionality.

Are there any alternatives other than writing `view` equivalent methods of the stateful methods(which shouldn't even be considered since static calls exist to avoid doing this, but this will also result in larger contract bytecode that may exceed the ~25KB limit)?

**Comment by shemnon** [comment]
The current support depends on running the EVM calls on the consensus nodes.  Because of the consensus interactions we can only support non-state-modifying calls.

We are working on code to bring these calls off of the consensus nodes to provide better `eth_estimateGas` calls. When that is ready we can remove the restriction.

**Comment by mshakeg** [comment]
@shemnon that's great to hear. #493 was opened relatively recently, so presumably work has only just begun to address that, do you have a rough estimate on when it'll be ready?

**Comment by shemnon** [comment]
I'm not in a position to comment publicly about delivery timeframes.

**Comment by mshakeg** [comment]
Hey @Nana-EC seems like little progress has been made on this issue in HIP-584, as attempting an `eth_call` on `SimpleStorage#setNumber` returns:

```json
{
  "code": -32008,
  "data": "ILLEGAL_STATE_CHANGE"
}
```

You can run the `hardhat-example` `test:ss` and `test:ss:previewnet` scripts on [my fork](https://github.com/mshakeg/hedera-json-rpc-relay) to demo the discrepancies between the hardhat node and the hedera previewnet.

**Comment by mshakeg** [comment]
Hey @Nana-EC not sure if Github notified you of my last comment. Could you please link the issue that tracks this feature? I couldn't find any related issue on the mirror node. I'm assuming there is as this falls within the scope of HIP-584.

**Comment by Nana-EC** [comment]
Hi @mshakeg HIP 584 development is in progress and we're working towards this.
We'll circle back with a ticket to track this or an update.

**Comment by Nana-EC** [comment]
Hey @mshakeg can you confirm your flows are addressed with the improvement in the 0.83 version of the mirror node which is in testnet. As Ivan K mentioned non system contract flows should work now for dynamic functions

**Comment by mshakeg** [comment]
Hey @Nana-EC I just confirmed that it works with my simple `test:ss` script. I'll re-open it if I get issues on my more complex use case.
